lib/common/tests/operations/parse_op_key.c

```diff
@@ -0,0 +1,226 @@
+    g_assert_cmpstr(rsc, ==, "vm_idcc_devel");
+    g_assert_cmpstr(ty, ==, "migrate_to");
+    g_assert_cmpint(ms, ==, 0);
+    free(rsc);
+    free(ty);
+}
+
+static void
+pre_post(void)
+{
+    char *rsc = NULL;
+    char *ty = NULL;
+    guint ms = 0;
+
+    g_assert_true(parse_op_key("rsc_drbd_7788:1_post_notify_start_0", &rsc, &ty, &ms));
+    g_assert_cmpstr(rsc, ==, "rsc_drbd_7788:1");
+    g_assert_cmpstr(ty, ==, "post_notify_start");
+    g_assert_cmpint(ms, ==, 0);
+    free(rsc);
+    free(ty);
+
+    g_assert_true(parse_op_key("rabbitmq-bundle-clone_pre_notify_stop_0", &rsc, &ty, &ms));
+    g_assert_cmpstr(rsc, ==, "rabbitmq-bundle-clone");
+    g_assert_cmpstr(ty, ==, "pre_notify_stop");
+    g_assert_cmpint(ms, ==, 0);
+    free(rsc);
+    free(ty);
+}
+
+static void
+skip_rsc(void)
+{
+    char *ty = NULL;
+    guint ms = 0;
+
+    g_assert_true(parse_op_key("Fencing_monitor_60000", NULL, &ty, &ms));
+    g_assert_cmpstr(ty, ==, "monitor");
+    g_assert_cmpint(ms, ==, 60000);
+    free(ty);
+}
+
+static void
+skip_ty(void)
+{
+    char *rsc = NULL;
+    guint ms = 0;
+
+    g_assert_true(parse_op_key("Fencing_monitor_60000", &rsc, NULL, &ms));
+    g_assert_cmpstr(rsc, ==, "Fencing");
+    g_assert_cmpint(ms, ==, 60000);
+    free(rsc);
+}
+
+static void
+skip_ms(void)
+{
+    char *rsc = NULL;
+    char *ty = NULL;
+
+    g_assert_true(parse_op_key("Fencing_monitor_60000", &rsc, &ty, NULL));
+    g_assert_cmpstr(rsc, ==, "Fencing");
+    g_assert_cmpstr(ty, ==, "monitor");
+    free(rsc);
+    free(ty);
+}
+
+static void
+empty_input(void)
+{
+    char *rsc = NULL;
+    char *ty = NULL;
+    guint ms = 0;
+
+    g_assert_false(parse_op_key("", &rsc, &ty, &ms));
+    g_assert_true(rsc == NULL);
+    g_assert_true(ty == NULL);
+    g_assert_cmpint(ms, ==, 0);
+
+    g_assert_false(parse_op_key(NULL, &rsc, &ty, &ms));
+    g_assert_true(rsc == NULL);
+    g_assert_true(ty == NULL);
+    g_assert_cmpint(ms, ==, 0);
+}
+
+static void
+malformed_input(void)
+{
+    char *rsc = NULL;
+    char *ty = NULL;
+    guint ms = 0;
+
+    g_assert_false(parse_op_key("httpd-bundle-0", &rsc, &ty, &ms));
+    g_assert(rsc == NULL);
+    g_assert(ty == NULL);
+    g_assert_cmpint(ms, ==, 0);
+
+    g_assert_false(parse_op_key("httpd-bundle-0_monitor", &rsc, &ty, &ms));
+    g_assert(rsc == NULL);
+    g_assert(ty == NULL);
+    g_assert_cmpint(ms, ==, 0);
+
+    g_assert_false(parse_op_key("httpd-bundle-0_30000", &rsc, &ty, &ms));
+    g_assert(rsc == NULL);
+    g_assert(ty == NULL);
+    g_assert_cmpint(ms, ==, 0);
+
+    /* FIXME: These should return FALSE but currently return TRUE.  Parsing
+     * is not smart enough to know the first doesn't have a resource name,
+     * nor that the second is missing a type.  Fixing those problems is
+     * probably difficult and not worth the effort.
+     */
+    /* g_assert_false(parse_op_key("post_notify_start_0", &rsc, &ty, &ms)); */
```

Comment:
Theoretically, this *could* be a start action for a resource named post_notify, so I think it's OK to successfully return post_notify + start + 0 in this case.